pkg/checks/latency.go

```diff
@@ -0,0 +1,171 @@
+package checks
+
+import (
+	"context"
+	"net/http"
+	"sync"
+	"time"
+
+	"github.com/getkin/kin-openapi/openapi3"
+	"github.com/mitchellh/mapstructure"
+	"golang.org/x/sync/errgroup"
+
+	"github.com/caas-team/sparrow/internal/logger"
+	"github.com/caas-team/sparrow/pkg/api"
+)
+
+var _ Check = (*Latency)(nil)
+
+func NewLatencyCheck() Check {
+	return &Latency{
+		mu:   sync.Mutex{},
+		cfg:  LatencyConfig{},
+		c:    nil,
+		done: make(chan bool),
+	}
+
+}
+
+type Latency struct {
+	cfg  LatencyConfig
+	mu   sync.Mutex
+	c    chan<- Result
+	done chan bool
+}
+
+type LatencyConfig struct {
+	Targets  []string
+	Interval time.Duration
+	Timeout  time.Duration
+}
+
+type LatencyResult struct {
+	Code  int     `json:"code"`
+	Error *string `json:"error"`
+	Total int64   `json:"total"`
+}
+
+func (l *Latency) Run(ctx context.Context) error {
+	log := logger.FromContext(ctx).WithGroup("Latency")
+	log.Info(l.cfg.Interval.String())
+	for {
+		select {
+		case <-ctx.Done():
+			log.Error("context canceled", "err", ctx.Err())
+			return ctx.Err()
+		case <-l.done:
+			return nil
+		case <-time.After(l.cfg.Interval):
+			results, err := l.check(ctx)
+			errval := ""
+			if err != nil {
+				errval = err.Error()
+			}
+			checkResult := Result{
+				Data:      results,
+				Err:       errval,
+				Timestamp: time.Now(),
+			}
+
+			l.c <- checkResult
+		}
+	}
+}
+
+func (l *Latency) Startup(ctx context.Context, cResult chan<- Result) error {
+	log := logger.FromContext(ctx).WithGroup("latency")
+	log.Debug("Starting latency check")
+
+	l.c = cResult
+	return nil
+}
+
+func (l *Latency) Shutdown(ctx context.Context) error {
+	l.done <- true
+	close(l.done)
+
+	return nil
+}
+
+func (l *Latency) SetConfig(ctx context.Context, config any) error {
+	var c LatencyConfig
+	err := mapstructure.Decode(config, &c)
+	if err != nil {
+		return ErrInvalidConfig
+	}
+	asSecond := time.Second * c.Interval
+	c.Interval = asSecond
+	l.mu.Lock()
+	defer l.mu.Unlock()
+	l.cfg = c
+
+	return nil
+}
+
+func (l *Latency) Schema() (*openapi3.SchemaRef, error) {
+	return OpenapiFromPerfData(make(map[string]LatencyResult))
+}
+
+func (l *Latency) RegisterHandler(ctx context.Context, router *api.RoutingTree) {
+	router.Add(http.MethodGet, "v1alpha1/latency", l.Handler)
+}
+
+func (l *Latency) DeregisterHandler(ctx context.Context, router *api.RoutingTree) {
+	router.Remove(http.MethodGet, "v1alpha1/latency")
+}
+
+func (l *Latency) Handler(w http.ResponseWriter, r *http.Request) {
+	w.WriteHeader(http.StatusOK)
+}
+
+func (l *Latency) check(ctx context.Context) (map[string]LatencyResult, error) {
+	log := logger.FromContext(ctx).WithGroup("check")
+	log.Debug("Checking latency")
+	var mu sync.Mutex
```

Comment:
Why not use the `Latency`'s Mutex? We don't really need to initiate a new mutex each time the function is called - this may lead to weird concurrency bugs, if the same `Latency` instance calls the `check` function in parallel goroutines.